File: src/main/scala/com/proyecto/api/auth/application/service/LoginService.scala
package com.proyecto.api.auth.application.service

import cats.effect.IO
import at.favre.lib.crypto.bcrypt.BCrypt
import com.proyecto.api.auth.domain.model.Credentials
import com.proyecto.api.auth.application.port.out.AuthUserRepository
import com.proyecto.api.auth.infrastructure.security.JwtService

class LoginService(userRepo: AuthUserRepository, jwtService: JwtService) {
  def login(credentials: Credentials): IO[Either[String, String]] = {
    userRepo.findByUsername(credentials.username).flatMap {
      case Some(user) =>
        IO.blocking(BCrypt.verifyer().verify(credentials.passwordRaw.toCharArray, user.passwordHash).verified).flatMap { isValid =>
          if (isValid) jwtService.generateToken(user.id).map(Right(_))
          else IO.pure(Left("Credenciales inválidas"))
        }
      case None =>
        IO.pure(Left("Credenciales inválidas")) // Mismo mensaje de error por seguridad
    }
  }
}

File: SphereChat.Api/Application/Services/LoginService.cs
using SphereChat.Api.Application.Ports.In;
using SphereChat.Api.Application.Ports.Out;
using SphereChat.Api.Domain.Models.Auth;

namespace SphereChat.Api.Application.Services;

/// <summary>
/// Servicio de Login funcional con BCrypt.
/// Equivalente a LoginService.scala (auth.application.service)
/// </summary>
public class LoginService
{
    private readonly IAuthUserRepository _userRepo;
    private readonly IJwtService _jwtService;

    public LoginService(IAuthUserRepository userRepo, IJwtService jwtService)
    {
        _userRepo = userRepo;
        _jwtService = jwtService;
    }

    public async Task<string?> LoginAsync(Credentials credentials)
    {
        var user = await _userRepo.FindByUsernameAsync(credentials.Username);
        if (user is null) return null;

        var isValid = BCrypt.Net.BCrypt.Verify(credentials.PasswordRaw, user.PasswordHash);
        if (!isValid) return null;

        return await _jwtService.GenerateTokenAsync(user.Id);
    }
}
